main.go

```diff
@@ -287,32 +300,63 @@ func newMarshalerType(fs *token.FileSet, imp types.Importer, typ *types.Named) *
 		if !f.Exported() {
 			continue
 		}
+		if f.Anonymous() {
+			fmt.Fprintf(os.Stderr, "Warning: ignoring embedded field %s\n", f.Name())
+			continue
+		}
+
 		mf := &marshalerField{
 			name:    f.Name(),
 			typ:     f.Type(),
 			origTyp: f.Type(),
 			tag:     styp.Tag(i),
 		}
-		if f.Anonymous() {
-			fmt.Fprintf(os.Stderr, "Warning: ignoring embedded field %s\n", f.Name())
-			continue
-		}
+
 		mtyp.Fields = append(mtyp.Fields, mf)
 	}
+
 	return mtyp
 }
 
+// findFunction returns a function with `name` that accepts no arguments
+// and returns a single value that is convertible to the given to type.
+func findFunction(typ *types.Named, name string, to types.Type) (*types.Func, types.Type) {
+	for i := 0; i < typ.NumMethods(); i++ {
+		fun := typ.Method(i)
+		if fun.Name() != name {
+			continue
+		}
+		if !fun.Exported() {
+			continue
+		}
+		sign := fun.Type().(*types.Signature)
+		if sign.Params().Len() != 0 || sign.Results().Len() != 1 {
+			continue
+		}
+		if err := checkConvertible(sign.Results().At(0).Type(), to); err == nil {
```

Comment:
I added a test to map/slice conv and it seems to work. Biggest drawback is that it will call the method multiple times which might not be optimal in case the function creates a copy each time.